### PR TITLE
EOS-16872 : LDRr2: Configure HAProxy as a part of S3 setup 3 Node VM component provisioning

### PR DESCRIPTION
**Pull request by shahidSTX**
EOS-16872 : LDRr2: Configure HAProxy as a part of S3 setup 3 Node VM component provisioning

Made fixes to add "s3 server port" and get public_ip from s3confstore.

Files modified :
	s3cortxutils/s3confstore/s3confstore/cortx_s3_confstore.py
	scripts/haproxy/s3haproxyconfig/s3haproxyconfig/s3_haproxy_config.py

Signed-off-by: Mohammed Shahid <mohammed.shahid@seagate.com>

### CODE REVIEW

File: s3cortxutils/s3confstore/s3confstore/cortx_s3_confstore.py [hunk 1]
```diff
@@ -225,6 +245,13 @@ def run(self):
       else:
         sys.exit("Failed to read private ip from confstore of machineid: {}".format(args.machineid))
 
+    elif args.command == 'getpublicip':
```
Comment (amitwac2608) [inline s3cortxutils/s3confstore/s3confstore/cortx_s3_confstore.py:248]: You need to add "getpublicip" to the 'args' first. Check above, line: 202

Also please test the code (both help and the option you are adding) should work fine. It seems you have not tested the "s3confstore binary"  !

**Review by amitwac2608 — COMMENTED** [review]
reviewed confstore changes.. Please ask Basav and Ajinkya to do haproxy file review

File: s3cortxutils/s3confstore/s3confstore/cortx_s3_confstore.py (hunk 1 above)
Comment (shahidSTX) [reply s3cortxutils/s3confstore/s3confstore/cortx_s3_confstore.py:251]: Addressed this and tested locally :
`[root@ssc-vm-2053 s3confstore]# s3confstore "json:///root/s3confstore.json" getpublicip --machineid 30512e5ae6df9f1ea02327bab45e499d
127.0.0.1
`

**Comment by shahidSTX** [comment]
Tested the changes locally by building s3confstore and s3haproxyconfig binaries :
`[root@ssc-vm-2053 s3confstore]# python3.6 setup.py install --force`
`...`
`Finished processing dependencies for s3confstore==2.0.0`
`[root@ssc-vm-2053 s3confstore]# s3confstore "json:///root/s3confstore.json" getpublicip --machineid 30512e5ae6df9f1ea02327bab45e499d`
`127.0.0.1`

`[root@ssc-vm-2053 s3haproxyconfig]# python3.6 setup.py install --force`
`...`
`Finished processing dependencies for s3haproxyconfig==2.0.0`
`[root@ssc-vm-2053 s3haproxyconfig]# s3haproxyconfig --path "json:///root/s3confstore.json"`
`[root@ssc-vm-2053 s3haproxyconfig]# systemctl status haproxy.service`
`● haproxy.service - HAProxy Load Balancer`
`   Loaded: loaded (/usr/lib/systemd/system/haproxy.service; disabled; vendor preset: disabled)`
`   Active: active (running) since Fri 2021-01-29 04:25:43 MST; 10s ago`
`  Process: 17666 ExecStartPre=/usr/sbin/haproxy -f $CONFIG -c -q (code=exited, status=0/SUCCESS)`

File: scripts/haproxy/s3haproxyconfig/s3haproxyconfig/s3_haproxy_config.py [hunk 2]
```diff
@@ -121,69 +139,55 @@ def run():
...
+
+    #Add complete information to haproxy.cfg file
+    cfg_file = '/etc/haproxy/haproxy.cfg'
+    target = open(cfg_file, "a+")
+
     target.write(header_text)
     target.write(str1)
+    target.write(
+        "    bind %s:80 ##### localhost 80 required for Auth - S3 connection\n"
+        "    #bind %s:443 ssl crt /etc/ssl/stx/stx.pem ### localhost required for CSM/UDX\n"
```
Comment (AjinkyaDhumal) [inline scripts/haproxy/s3haproxyconfig/s3haproxyconfig/s3_haproxy_config.py:157]: Uncomment these 3 lines of ssl

Comment (shahidSTX) [reply scripts/haproxy/s3haproxyconfig/s3haproxyconfig/s3_haproxy_config.py]: Done!

**Comment by shahidSTX** [comment]
After copying stx.pem file from  ansible directory, testing has passed for s3haproxyconfig binary :
`[root@ssc-vm-2053 s3haproxyconfig]# systemctl status haproxy.service`
`● haproxy.service - HAProxy Load Balancer`
`   Loaded: loaded (/usr/lib/systemd/system/haproxy.service; disabled; vendor preset: disabled)`
`   Active: active (running) since Mon 2021-02-01 03:05:29 MST; 11s ago`
`  Process: 109772 ExecStartPre=/usr/sbin/haproxy -f $CONFIG -c -q (code=exited, status=0/SUCCESS)`
` Main PID: 109775 (haproxy)`
`   CGroup: /system.slice/haproxy.service`
`├─109775 /usr/sbin/haproxy -Ws -f /etc/haproxy/haproxy.cfg -p /run/haproxy.pid`
`└─109776 /usr/sbin/haproxy -Ws -f /etc/haproxy/haproxy.cfg -p /run/haproxy.pid`

**Comment by shahidSTX** [comment]
s3iamcli basic cmds are working with 2 set of pem files, I tried one round of testing with each of the files : `cortx-s3server/scripts/haproxy/ssl/s3.seagate.com.pem` and `cortx-s3server/ansible/files/certs/stx-s3/s3/s3server.pem` and below was the o/p in both the cases :
`[root@ssc-vm-2053 s3server-main]# s3iamcli listaccounts --no-ssl`
`Enter Ldap User Id: sgiamadmin`
`Enter Ldap password:`
`No accounts found.`

**Review by AjinkyaDhumal — APPROVED** [review]

File: s3cortxutils/s3confstore/s3confstore/cortx_s3_confstore.py [hunk 3]
```diff
@@ -124,6 +124,26 @@ def get_privateip(self, machine_id: str):
 
     return privateip
 
+  def get_publicip(self, machine_id: str):
```
Comment (amitwac2608) [inline s3cortxutils/s3confstore/s3confstore/cortx_s3_confstore.py:127]: please add UT for this function.

**Review by amitwac2608 — COMMENTED** [review]
checked the confstore file change..

File: scripts/haproxy/s3haproxyconfig/s3haproxyconfig/s3_haproxy_config.py [hunk 4]
```diff
@@ -58,26 +87,15 @@ def run():
...
-    numS3Instances = int(s3conf_store.get_s3instance_count(machine_id))
-
-    if numS3Instances <= 0:
-        numS3Instances = 1
-
-    s3inport = 28081
-    s3auport = 28050
-
+    #Set the string literals to be added to config file
     str1 = '''
```
Comment (bkirunge7) [inline scripts/haproxy/s3haproxyconfig/s3haproxyconfig/s3_haproxy_config.py:91]: str1, str2 rename these variables with proper name like front_end_header_text

File: s3cortxutils/s3confstore/s3confstore/cortx_s3_confstore.py (hunk 3 above)
Comment (shahidSTX) [reply s3cortxutils/s3confstore/s3confstore/cortx_s3_confstore.py:127]: Done.

File: scripts/haproxy/s3haproxyconfig/s3haproxyconfig/s3_haproxy_config.py (hunk 4 above)
Comment (shahidSTX) [reply scripts/haproxy/s3haproxyconfig/s3haproxyconfig/s3_haproxy_config.py]: Done

**Comment by shahidSTX** [comment]
Successful executed all earlier testing and also did UT for 3confstore : 
### 
[root@ssc-vm-2053 s3server-main]# ./s3cortxutils/s3confstore/scripts/run_all_ut.sh

Running s3confstore UT's...
=================================================================================== test session starts ====================================================================================
platform linux -- Python 3.6.8, pytest-2.9.2, py-1.4.32, pluggy-0.3.1 -- /usr/bin/python36
cachedir: s3cortxutils/s3confstore/.cache
rootdir: /root/workspace/s3server-main/s3cortxutils/s3confstore, inifile:
collected 19 items

s3cortxutils/s3confstore/ut/s3confstore_unit_tests.py::S3ConfStoreAPIsUT::test_get_nodecount_none PASSED
s3cortxutils/s3confstore/ut/s3confstore_unit_tests.py::S3ConfStoreAPIsUT::test_get_nodecount_success PASSED
s3cortxutils/s3confstore/ut/s3confstore_unit_tests.py::S3ConfStoreAPIsUT::test_get_nodenames_list_empty PASSED
s3cortxutils/s3confstore/ut/s3confstore_unit_tests.py::S3ConfStoreAPIsUT::test_get_nodenames_list_exception PASSED
s3cortxutils/s3confstore/ut/s3confstore_unit_tests.py::S3ConfStoreAPIsUT::test_get_nodenames_list_success PASSED
s3cortxutils/s3confstore/ut/s3confstore_unit_tests.py::S3ConfStoreAPIsUT::test_get_privateip_emptystring PASSED
s3cortxutils/s3confstore/ut/s3confstore_unit_tests.py::S3ConfStoreAPIsUT::test_get_privateip_success PASSED
s3cortxutils/s3confstore/ut/s3confstore_unit_tests.py::S3ConfStoreAPIsUT::test_get_publicip_check PASSED
s3cortxutils/s3confstore/ut/s3confstore_unit_tests.py::S3ConfStoreAPIsUT::test_get_publicip_emptystring PASSED
s3cortxutils/s3confstore/ut/s3confstore_unit_tests.py::S3ConfStoreAPIsUT::test_get_publicip_success PASSED
s3cortxutils/s3confstore/ut/s3confstore_unit_tests.py::S3ConfStoreAPIsUT::test_get_s3instancecount_success PASSED
s3cortxutils/s3confstore/ut/s3confstore_unit_tests.py::S3ConfStoreAPIsUT::test_json_conf PASSED
s3cortxutils/s3confstore/ut/s3confstore_unit_tests.py::S3ConfStoreAPIsUT::test_mock_get PASSED
s3cortxutils/s3confstore/ut/s3confstore_unit_tests.py::S3ConfStoreAPIsUT::test_mock_load PASSED
s3cortxutils/s3confstore/ut/s3confstore_unit_tests.py::S3ConfStoreAPIsUT::test_mock_save PASSED
s3cortxutils/s3confstore/ut/s3confstore_unit_tests.py::S3ConfStoreAPIsUT::test_mock_set PASSED
s3cortxutils/s3confstore/ut/s3confstore_unit_tests.py::S3ConfStoreAPIsUT::test_validate_configfile_doesnotexist PASSED
s3cortxutils/s3confstore/ut/s3confstore_unit_tests.py::S3ConfStoreAPIsUT::test_validate_configfile_unsupportedformat PASSED
s3cortxutils/s3confstore/ut/s3confstore_unit_tests.py::S3ConfStoreAPIsUT::test_yaml_conf PASSED

================================================================================ 19 passed in 0.23 seconds =================================================================================
s3confstore UT's runs successfully
###

**Review by amitwac2608 — APPROVED** [review]
looks good

**Comment by shahidSTX** [comment]
Manual jenkins job passed : http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/S3server-7.8.2003/536/

**Comment by shahidSTX** [comment]
pre-merge job passed : http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Component-Test/job/S3Server-PreMerge-Test/65/

mini-provisioning passed : http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Mini-Provisioning/job/cortx-s3server/150/

**Pull request merged**